src/corecel/data/LdgIterator.hh

```diff
@@ -0,0 +1,261 @@
+//---------------------------------*-CUDA-*----------------------------------//
+// Copyright 2023 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file corecel/data/LdgIterator.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include <cstddef>
+#include <iterator>
+#include <type_traits>
+
+#include "corecel/Macros.hh"
+#include "corecel/Types.hh"
+#include "corecel/math/Algorithms.hh"
+
+namespace celeritas
+{
+namespace detail
+{
+//---------------------------------------------------------------------------//
+
+template<class T, typename = void>
+struct LDGLoad
+{
+    using value_type = T;
+    using pointer = std::add_pointer_t<value_type const>;
+    using reference = std::add_lvalue_reference_t<value_type const>;
```

Comment:
since `__ldg` should only ever return constant primitive type it should be cheaper to return a copy instead of const ref in most cases?

Reply:
I would think the compiler would optmize to the same thing if it reutrns an lvalue or not... but isn't the signature `T __ldg(T*)` i.e. it always returns an rvalue?

Reply:
yeah, a function call that returns a non-reference is rvalue. In that case:
```c++
T const& f(T const* p)
{
    return __ldg(p);
}
```
I'd assume we bind a reference to temporary which results in UB?